### PR TITLE
bpo-32107 - Fix test_uuid

### PR DESCRIPTION
**Pull request by warsaw**
The original check was erroneous in two ways.  First, the documentation said
that "47 bit will never be set in IEEE 802 addresses obtained from network
cards".  I think this is referring to the universally vs. locally administered
MAC addresses.  Network cards from hardware manufacturers will always be
universally administered in order to guarantee global uniqueness.

But from https://en.wikipedia.org/wiki/MAC_address it says that this flag is
"distinguished by setting the second-least-significant bit of the first
octet of the address", where a 0 means universally administered and a 1 means
it's locally administered.  This works out to the 42nd bit when counting the
LSB as bit 1, or 1<<41.

The second bug is that the original bitmask value isn't right for either
description:

% ./python.exe -c "from math import log2; print(log2(0x010000000000))"
40.0

This causes the test to fail on valid MAC addresses.

Fix this by improving the comment, with references, and using the correct
bitmask.

https://bugs.python.org/issue32107

### CODE REVIEW

**Comment by warsaw** [comment]
I also added a very minor simplification to the code, since it's always better to return an explicit value when there are other return paths in the code, rather than letting None be implicitly returned.

File: Lib/uuid.py [hunk 1]
```diff
@@ -404,8 +404,7 @@ def _arp_getnode():
     # This works on Linux, FreeBSD and NetBSD
     mac = _find_mac('arp', '-an', [os.fsencode('(%s)' % ip_addr)],
                     lambda i: i+2)
-    if mac:
-        return mac
+    return mac
```
Comment (serhiy-storchaka) [inline Lib/uuid.py]: Should return None instead of 0.

Comment (warsaw) [reply Lib/uuid.py]: Ah, good point.  Let me fix that.

File: Lib/uuid.py [hunk 2]
```diff
@@ -404,8 +404,8 @@ def _arp_getnode():
     # This works on Linux, FreeBSD and NetBSD
     mac = _find_mac('arp', '-an', [os.fsencode('(%s)' % ip_addr)],
                     lambda i: i+2)
-    if mac:
-        return mac
+    # Return None instead of 0.
+    return mac if mac else None
```
Comment (serhiy-storchaka) [inline Lib/uuid.py]: And if pep8-izy the code, I suggest to add explicit `return None` in other functions.

For uniformity I would prefer
```python
    if mac:
        return mac
    return None
```
in this function.

Comment (vstinner) [reply Lib/uuid.py]: I like the code @serhiy-storchaka proposed.

Comment (warsaw) [reply Lib/uuid.py]: @vstinner Yep, agreed.

**Comment by warsaw** [comment]
@serhiy-storchaka @vstinner - I think this one's got it.  The functions should never return a locally administered MAC address now.

ED: Ah darn maybe not.  Too tired.  I'll get back to this later.

File: Lib/uuid.py [hunk 3]
```diff
@@ -355,7 +370,7 @@ def _find_mac(command, args, hw_identifiers, get_index):
                         try:
                             word = words[get_index(i)]
                             mac = int(word.replace(b':', b''), 16)
-                            if mac:
+                            if ~(mac & (1<<41)):
```
Comment (vstinner) [inline Lib/uuid.py]: I don't understand the usage of the ~int operator. 

What do you want to test? If a bit is unset? Do you mean "if not(mac & (1<<41)):" or "if (mac & (1<<41)) == 0:" ?

Please write a private subfunction for this test.

**Review by vstinner — COMMENTED** [review]
I understand that you want to exclude "local MAC". Ok. But what if a compute only has local MACs? uuid1() fails with a TypeError with a strange error message if getnode() returns None. At least, please modify uuid1() to raise a better error if no MAC can be found.

Or maybe fall back on a local MAC if no other address can be found?

File: Lib/uuid.py (hunk 3 above)
Comment (warsaw) [reply Lib/uuid.py]: Yeah, that was completely botched.  Now that I've had some sleep <wink> I've fixed this with a helper function.

Comment (warsaw) [reply Lib/uuid.py]: Also with the latest commit, we keep track of the first locally administered MAC found.  If no universally administered MAC is found, we return the first local MAC.

And the way the getters work, if there are no local or universal MACs found, then `getnode()` ultimately falls back to the RFC 4122 specified random MAC, with the multicast bit set.  So it should never be the case that `getnode()` doesn't return an RFC-defined node address.

**Comment by warsaw** [comment]
```
>>> mac = 0x0242ac11000d
>>> mac & (1<<41)
2199023255552
```

Which means that travis is returning a locally administered MAC, which must mean it has no universally administered MACs.   This is what I was alluding to previously about environmental factors leaking into the tests.

**Comment by warsaw** [comment]
As the last commit's debugging output shows, there are only local interfaces on Travis.

**Comment by warsaw** [comment]
I think there's nothing better that we can do except skip the two failing tests under Travis.  This last commit does that, so we'll see.

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by warsaw** [comment]
Oh hells bells, the merge went badly.

**Comment by warsaw** [comment]
See #4576 which resolves the merge snafus.

**Pull request closed without merging**